### PR TITLE
Fix crash in explain graph with StorageMerge

### LINKED ISSUE #48053: Segfault/heap-use-after-free in `printPipelineCompact`
https://s3.amazonaws.com/clickhouse-test-reports/0/bc2b6257f2a3c3b1534448fd872c216fd59602a0/fuzzer_astfuzzertsan/report.html

```
WARNING: ThreadSanitizer: heap-use-after-free (virtual call vs free) (pid=165)
  Read of size 8 at 0x7b70005e2800 by thread T3:
    #0 DB::printPipelineCompact(std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>> const&, DB::WriteBuffer&, bool) build_docker/../src/QueryPipeline/printPipeline.cpp:116:53 (clickhouse+0x1d91848a) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #1 DB::InterpreterExplainQuery::executeImpl() build_docker/../src/Interpreters/InterpreterExplainQuery.cpp:508:25 (clickhouse+0x1d816e4f) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #2 DB::InterpreterExplainQuery::execute() build_docker/../src/Interpreters/InterpreterExplainQuery.cpp:87:20 (clickhouse+0x1d813057) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #3 DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/../src/Interpreters/executeQuery.cpp:713:36 (clickhouse+0x1de42226) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #4 DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) build_docker/../src/Interpreters/executeQuery.cpp:1160:30 (clickhouse+0x1de3e98b) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #5 DB::TCPHandler::runImpl() build_docker/../src/Server/TCPHandler.cpp:420:24 (clickhouse+0x1f18da2a) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #6 DB::TCPHandler::run() build_docker/../src/Server/TCPHandler.cpp:2037:9 (clickhouse+0x1f1a37a7) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #7 Poco::Net::TCPServerConnection::start() build_docker/../base/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x22f25762) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #8 Poco::Net::TCPServerDispatcher::run() build_docker/../base/poco/Net/src/TCPServerDispatcher.cpp:115:20 (clickhouse+0x22f25fd2) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #9 Poco::PooledThread::run() build_docker/../base/poco/Foundation/src/ThreadPool.cpp:188:14 (clickhouse+0x231803f9) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #10 Poco::(anonymous namespace)::RunnableHolder::run() build_docker/../base/poco/Foundation/src/Thread.cpp:45:11 (clickhouse+0x2317e7af) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #11 Poco::ThreadImpl::runnableEntry(void*) build_docker/../base/poco/Foundation/src/Thread_POSIX.cpp:335:27 (clickhouse+0x2317cdc7) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)

  Previous write of size 8 at 0x7b70005e2800 by thread T3:
    #0 operator delete(void*, unsigned long) <null> (clickhouse+0xc41c3de) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #1 DB::ReadFromMergeTree::~ReadFromMergeTree() build_docker/../src/Processors/QueryPlan/ReadFromMergeTree.h:32:7 (clickhouse+0x1f75b826) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #2 std::__1::default_delete<DB::IQueryPlanStep>::operator()[abi:v15000](DB::IQueryPlanStep*) const build_docker/../contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48:5 (clickhouse+0x1f729397) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #3 std::__1::unique_ptr<DB::IQueryPlanStep, std::__1::default_delete<DB::IQueryPlanStep>>::reset[abi:v15000](DB::IQueryPlanStep*) build_docker/../contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305:7 (clickhouse+0x1f729397)
    #4 std::__1::unique_ptr<DB::IQueryPlanStep, std::__1::default_delete<DB::IQueryPlanStep>>::~unique_ptr[abi:v15000]() build_docker/../contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259:19 (clickhouse+0x1f729397)
    #5 DB::QueryPlan::Node::~Node() build_docker/../src/Processors/QueryPlan/QueryPlan.h:102:12 (clickhouse+0x1f729397)
    #6 void std::__1::__destroy_at[abi:v15000]<DB::QueryPlan::Node, 0>(DB::QueryPlan::Node*) build_docker/../contrib/llvm-project/libcxx/include/__memory/construct_at.h:63:13 (clickhouse+0x1f729397)
    #7 void std::__1::destroy_at[abi:v15000]<DB::QueryPlan::Node, 0>(DB::QueryPlan::Node*) build_docker/../contrib/llvm-project/libcxx/include/__memory/construct_at.h:88:5 (clickhouse+0x1f729397)
    #8 void std::__1::allocator_traits<std::__1::allocator<std::__1::__list_node<DB::QueryPlan::Node, void*>>>::destroy[abi:v15000]<DB::QueryPlan::Node, void, void>(std::__1::allocator<std::__1::__list_node<DB::QueryPlan::Node, void*>>&, DB::QueryPlan::Node*) build_docker/../contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:317:9 (clickhouse+0x1f729397)
    #9 std::__1::__list_imp<DB::QueryPlan::Node, std::__1::allocator<DB::QueryPlan::Node>>::clear() build_docker/../contrib/llvm-project/libcxx/include/list:743:13 (clickhouse+0x1f729397)
    #10 std::__1::__list_imp<DB::QueryPlan::Node, std::__1::allocator<DB::QueryPlan::Node>>::~__list_imp() build_docker/../contrib/llvm-project/libcxx/include/list:724:3 (clickhouse+0x1f729397)
    #11 DB::QueryPlan::~QueryPlan() build_docker/../src/Processors/QueryPlan/QueryPlan.cpp:30:23 (clickhouse+0x1f729397)
    #12 DB::ReadFromMerge::createSources(std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, DB::QueryProcessingStage::Enum const&, unsigned long, DB::Block const&, std::__1::vector<DB::ReadFromMerge::AliasData, std::__1::allocator<DB::ReadFromMerge::AliasData>> const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IStorage>, std::__1::shared_ptr<DB::RWLockImpl::LockHolderImpl>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>, std::__1::shared_ptr<DB::Context>, unsigned long, bool) build_docker/../src/Storages/StorageMerge.cpp:703:5 (clickhouse+0x1e59c039) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #13 DB::ReadFromMerge::initializePipeline(DB::QueryPipelineBuilder&, DB::BuildQueryPipelineSettings const&) build_docker/../src/Storages/StorageMerge.cpp:518:32 (clickhouse+0x1e597b12) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #14 DB::ISourceStep::updatePipeline(std::__1::vector<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>, std::__1::allocator<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>>>, DB::BuildQueryPipelineSettings const&) build_docker/../src/Processors/QueryPlan/ISourceStep.cpp:20:5 (clickhouse+0x1f707915) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #15 DB::QueryPlan::buildQueryPipeline(DB::QueryPlanOptimizationSettings const&, DB::BuildQueryPipelineSettings const&) build_docker/../src/Processors/QueryPlan/QueryPlan.cpp:187:47 (clickhouse+0x1f72ae9b) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #16 DB::InterpreterExplainQuery::executeImpl() build_docker/../src/Interpreters/InterpreterExplainQuery.cpp:496:38 (clickhouse+0x1d816cd2) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #17 DB::InterpreterExplainQuery::execute() build_docker/../src/Interpreters/InterpreterExplainQuery.cpp:87:20 (clickhouse+0x1d813057) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #18 DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/../src/Interpreters/executeQuery.cpp:713:36 (clickhouse+0x1de42226) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #19 DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) build_docker/../src/Interpreters/executeQuery.cpp:1160:30 (clickhouse+0x1de3e98b) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #20 DB::TCPHandler::runImpl() build_docker/../src/Server/TCPHandler.cpp:420:24 (clickhouse+0x1f18da2a) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #21 DB::TCPHandler::run() build_docker/../src/Server/TCPHandler.cpp:2037:9 (clickhouse+0x1f1a37a7) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #22 Poco::Net::TCPServerConnection::start() build_docker/../base/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x22f25762) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #23 Poco::Net::TCPServerDispatcher::run() build_docker/../base/poco/Net/src/TCPServerDispatcher.cpp:115:20 (clickhouse+0x22f25fd2) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #24 Poco::PooledThread::run() build_docker/../base/poco/Foundation/src/ThreadPool.cpp:188:14 (clickhouse+0x231803f9) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #25 Poco::(anonymous namespace)::RunnableHolder::run() build_docker/../base/poco/Foundation/src/Thread.cpp:45:11 (clickhouse+0x2317e7af) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #26 Poco::ThreadImpl::runnableEntry(void*) build_docker/../base/poco/Foundation/src/Thread_POSIX.cpp:335:27 (clickhouse+0x2317cdc7) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)

  Thread T3 'TCPHandler' (tid=170, running) created by main thread at:
    #0 pthread_create <null> (clickhouse+0xc39c0bd) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #1 Poco::ThreadImpl::startImpl(Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>) build_docker/../base/poco/Foundation/src/Thread_POSIX.cpp:198:6 (clickhouse+0x2317c7d3) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #2 Poco::Thread::start(Poco::Runnable&) build_docker/../base/poco/Foundation/src/Thread.cpp:118:2 (clickhouse+0x2317e1ec) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #3 Poco::PooledThread::start() build_docker/../base/poco/Foundation/src/ThreadPool.cpp:78:10 (clickhouse+0x231808f0) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #4 Poco::ThreadPool::ThreadPool(int, int, int, int) build_docker/../base/poco/Foundation/src/ThreadPool.cpp:237:12 (clickhouse+0x231808f0)
    #5 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/../programs/server/Server.cpp:795:22 (clickhouse+0x144cc3f4) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #6 Poco::Util::Application::run() build_docker/../base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x22f3e69d) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #7 DB::Server::run() build_docker/../programs/server/Server.cpp:476:25 (clickhouse+0x144c739c) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #8 Poco::Util::ServerApplication::run(int, char**) build_docker/../base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x22f5e30f) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #9 mainEntryClickHouseServer(int, char**) build_docker/../programs/server/Server.cpp:197:20 (clickhouse+0x144c3ed4) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)
    #10 main build_docker/../programs/main.cpp:481:12 (clickhouse+0xc41d7d1) (BuildId: 29238f892003f85b722a003a6d999d30d05d2b77)

SUMMARY: ThreadSanitizer: heap-use-after-free (virtual call vs free) build_docker/../src/QueryPipeline/printPipeline.cpp:116:53 in DB::printPipelineCompact(std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>> const&, DB::WriteBuffer&, bool)

```

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Fix crash in `EXPLAIN graph = 1 ...` with StorageMerge, close https://github.com/ClickHouse/ClickHouse/issues/48053

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by evillique — APPROVED** [review]

File: src/Storages/StorageMerge.h [hunk 1]
```diff
@@ -159,6 +159,8 @@ class ReadFromMerge final : public SourceStepWithFilter
     StoragePtr storage_merge;
     StorageSnapshotPtr merge_storage_snapshot;
 
+    std::vector<QueryPlan> child_plans;
```
Comment (KochetovNicolai) [inline src/Storages/StorageMerge.h:162]: No, it's not a good idea.
Storage may leave for a long time, but we don't clean these plans.

**Review by KochetovNicolai — APPROVED** [review]

File: src/Storages/StorageMerge.cpp [hunk 2]
```diff
@@ -655,7 +655,7 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
         if (real_column_names.empty())
             real_column_names.push_back(ExpressionActions::getSmallestColumn(storage_snapshot->metadata->getColumns().getAllPhysical()).name);
 
-        QueryPlan plan;
+        QueryPlan & plan = child_plans.emplace_back();
```
Comment (alexey-milovidov) [inline src/Storages/StorageMerge.cpp:658]: Missing a comment.

File: src/Storages/StorageMerge.h (hunk 1 above)
Comment (alexey-milovidov) [inline src/Storages/StorageMerge.h:162]: Missing a comment.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**